### PR TITLE
docs: translate 'Rules of React'

### LINKED ISSUE #651: Rules of React
Page: [id.react.dev/reference/rules](https://id.react.dev/reference/rules)

Assignee:

### PR DESCRIPTION
**Pull request by ceavinrufus**
Closes #651

## Description

Translate the Rules of React page.
Page URL: [id.react.dev/reference/rules](https://id.react.dev/reference/rules)

### CODE REVIEW

**Comment by r17x** [comment]
Sorry @ceavinrufus it duplicated with @resir014 PR https://github.com/reactjs/id.react.dev/pull/657

File: src/content/reference/rules/index.md [hunk 1]
```diff
@@ -1,52 +1,52 @@
 ---
-title: Rules of React
+title: Aturan React
```
Comment (r17x) [inline src/content/reference/rules/index.md:2]:
```suggestion
title: Peraturan React
```

File: src/content/reference/rules/index.md [hunk 2]
```diff
@@ -1,52 +1,52 @@
...
 
 ---
 
 <Note>
-To learn more about expressing UIs with React, we recommend reading [Thinking in React](/learn/thinking-in-react).
+Untuk mempelajari lebih lanjut tentang mengekspresikan antarmuka dengan React, kami sarankan untuk membaca [Cara Berpikir dengan React](/learn/thinking-in-react).
 </Note>
 
-This section describes the rules you need to follow to write idiomatic React code. Writing idiomatic React code can help you write well organized, safe, and composable applications. These properties make your app more resilient to changes and makes it easier to work with other developers, libraries, and tools.
+Bagian ini menjelaskan aturan yang perlu Anda ikuti untuk menulis kode React yang idiomatis. Menulis kode React yang idiomatis dapat membantu Anda menulis aplikasi yang terorganisisasi dengan baik, aman, dan dapat disusun. Hal ini membuat aplikasi Anda lebih tahan terhadap perubahan dan memudahkan bekerja dengan pengembang, pustaka (_library_), dan alat lain.
```
Comment (r17x) [inline src/content/reference/rules/index.md:17]: Use asterisk `*` instead underscore `_`
```suggestion
Bagian ini menjelaskan aturan yang perlu Anda ikuti untuk menulis kode React yang idiomatis. Menulis kode React yang idiomatis dapat membantu Anda menulis aplikasi yang terorganisisasi dengan baik, aman, dan dapat disusun. Hal ini membuat aplikasi Anda lebih tahan terhadap perubahan dan memudahkan bekerja dengan pengembang, pustaka (*library*), dan alat lain.
```

File: src/content/reference/rules/index.md [hunk 3]
```diff
@@ -1,52 +1,52 @@
...
 
-These rules are known as the **Rules of React**. They are rules – and not just guidelines – in the sense that if they are broken, your app likely has bugs. Your code also becomes unidiomatic and harder to understand and reason about.
```
Comment (r17x) [inline src/content/reference/rules/index.md:19]: Use asterisk * instead underscore _

File: src/content/reference/rules/index.md [hunk 4]
```diff
@@ -1,52 +1,52 @@
...
+Aturan-aturan ini dikenal sebagai **Aturan React**. Ini adalah aturan, bukan sekadar pedoman, yang dalam artian jika aturan tersebut dilanggar, kemungkinan besar aplikasi Anda memiliki _bug_. Kode Anda juga menjadi tidak idiomatis dan lebih sulit untuk dipahami dan dipikirkan.
 
-We strongly recommend using [Strict Mode](/reference/react/StrictMode) alongside React's [ESLint plugin](https://www.npmjs.com/package/eslint-plugin-react-hooks) to help your codebase follow the Rules of React. By following the Rules of React, you'll be able to find and address these bugs and keep your application maintainable.
+Kami sangat merekomendasikan penggunaan [Strict Mode](/reference/react/StrictMode) bersama dengan [_plugin_ ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks) untuk React, untuk membantu basis kode (codebase) Anda mematuhi Aturan React. Dengan mengikuti Aturan React, Anda akan dapat menemukan dan mengatasi bug ini dan menjaga aplikasi Anda tetap dapat dipelihara.
```
Comment (r17x) [inline src/content/reference/rules/index.md:21]: * Use asterisk * instead underscore _
* Use asterisk * in "english" terms (e.g: codebase) or use bahasa `basis kode`

File: src/content/reference/rules/index.md [hunk 5]
```diff
@@ -1,52 +1,52 @@
...
 
 ---
 
-## Components and Hooks must be pure {/*components-and-hooks-must-be-pure*/}
+## Komponen dan _Hooks_ harus murni {/*components-and-hooks-must-be-pure*/}
```
Comment (r17x) [inline src/content/reference/rules/index.md:25]: Use asterisk * instead underscore _

File: src/content/reference/rules/index.md [hunk 6]
```diff
@@ -1,52 +1,52 @@
...
 
-[Purity in Components and Hooks](/reference/rules/components-and-hooks-must-be-pure) is a key rule of React that makes your app predictable, easy to debug, and allows React to automatically optimize your code.
+[Kemurnian dalam Komponen dan _Hooks_](/reference/rules/components-and-hooks-must-be-pure) adalah aturan utama React yang membuat aplikasi Anda dapat diprediksi, mudah di-_debug_, dan memungkinkan React mengoptimalkan kode Anda secara otomatis.
```
Comment (r17x) [inline src/content/reference/rules/index.md:27]: Use asterisk * instead underscore _

File: src/content/reference/rules/index.md [hunk 7]
```diff
@@ -1,52 +1,52 @@
...
+* [Efek samping harus dijalankan di luar _render_](/reference/rules/components-and-hooks-must-be-pure#side-effects-must-run-outside-of-render) – Efek samping tidak boleh dijalankan dalam _render_, karena React dapat me-_render_ komponen beberapa kali untuk menciptakan pengalaman pengguna sebaik mungkin.
+* [_Props_ dan _state_ adalah sesuatu yang _immutable_](/reference/rules/components-and-hooks-must-be-pure#props-and-state-are-immutable) – _Props_ dan _state_ suatu komponen adalah _snapshot_ yang _immutable_ untuk sekali _render_. Jangan pernah mengubahnya secara langsung.
+* [Nilai kembalian dan argumen ke _Hooks_ adalah sesuatu yang _immutable_](/reference/rules/components-and-hooks-must-be-pure#return-values-and-arguments-to-hooks-are-immutable) – Setelah nilai dioper ke sebuah _Hook_, Anda tidak boleh mengubahnya. Seperti _props_ dalam JSX, nilai menjadi _immutable_ saat dioper ke _Hook_.
+* [Nilai menjadi _immutable_ setelah dioper ke JSX](/reference/rules/components-and-hooks-must-be-pure#values-are-immutable-after-being-passed-to-jsx) – Jangan ubah nilai setelah digunakan di JSX. Pindahkan perubahan (_mutation_) sebelum JSX dibuat.
```
Comment (r17x) [inline src/content/reference/rules/index.md:33]: Use asterisk * instead underscore _

File: src/content/reference/rules/index.md [hunk 8]
```diff
@@ -1,52 +1,52 @@
...
 
-* [Only call Hooks at the top level](/reference/rules/rules-of-hooks#only-call-hooks-at-the-top-level) – Don’t call Hooks inside loops, conditions, or nested functions. Instead, always use Hooks at the top level of your React function, before any early returns.
-* [Only call Hooks from React functions](/reference/rules/rules-of-hooks#only-call-hooks-from-react-functions) – Don’t call Hooks from regular JavaScript functions.
+* [Panggil _Hooks_ hanya di tingkat teratas](/reference/rules/rules-of-hooks#only-call-hooks-at-the-top-level) – Jangan panggil _Hooks_ di dalam perulangan (_loop_), kondisi (_if/else_), atau fungsi bersarang. Sebagai gantinya, selalu gunakan _Hooks_ di level paling atas fungsi React Anda, sebelum kembalian awal (jika ada).
+* [Panggil _Hooks_ hanya dari fungsi React](/reference/rules/rules-of-hooks#only-call-hooks-from-react-functions) – Jangan panggil _Hooks_ dari fungsi JavaScript biasa.
```
Comment (r17x) [inline src/content/reference/rules/index.md:51]: * Use asterisk * instead underscore _
* Use asterisk * in "english" terms (e.g: codebase) or use bahasa basis kode

**Review by r17x — CHANGES_REQUESTED** [review]

**Comment by ceavinrufus** [comment]
Could you kindly review my code again, please? @r17x

File: src/content/reference/rules/index.md [hunk 9]
```diff
@@ -1,52 +1,52 @@
...
+* [Jangan pernah memanggil fungsi komponen secara langsung](/reference/rules/react-calls-components-and-hooks#never-call-component-functions-directly) – Komponen hanya boleh digunakan di JSX. Jangan panggil mereka sebagai fungsi biasa.
+* [Jangan pernah mengoper *hooks* sebagai nilai biasa](/reference/rules/react-calls-components-and-hooks#never-pass-around-hooks-as-regular-values) – *Hooks* hanya boleh dipanggil di dalam komponen. Jangan pernah mengopernya sebagai nilai biasa.
 
 ---
 
-## Rules of Hooks {/*rules-of-hooks*/}
+## Peraturan *Hooks* {/*rules-of-hooks*/}
 
-Hooks are defined using JavaScript functions, but they represent a special type of reusable UI logic with restrictions on where they can be called. You need to follow the [Rules of Hooks](/reference/rules/rules-of-hooks) when using them.
+Meskipun *Hooks* ditulis menggunakan fungsi JavaScript, mereka memiliki fungsi khusus sebagai logika antarmuka yang dapat digunakan kembali. Namun, ada batasan terkait saat Anda bisa menggunakan Hook. Anda perlu mematuhi [Peraturan *Hooks*](/reference/rules/rules-of-hooks) saat menggunakannya.
```
Comment (r17x) [inline src/content/reference/rules/index.md:48]:
```suggestion
Meskipun *Hooks* ditulis menggunakan fungsi JavaScript, mereka memiliki fungsi khusus sebagai logika antarmuka yang dapat digunakan kembali. Namun, ada batasan terkait saat Anda bisa menggunakan *Hook*. Anda perlu mematuhi [Peraturan *Hooks*](/reference/rules/rules-of-hooks) saat menggunakannya.
```

**Review by r17x — CHANGES_REQUESTED** [review]

**Review by r17x — APPROVED** [review]

**Review by resir014 — APPROVED** [review]

**Pull request merged**